api_server.go

```diff
@@ -0,0 +1,196 @@
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"github.com/sirupsen/logrus"
+	"html/template"
+	"net"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+	"time"
+)
+
+type ApiServer struct {
+	ObjectStore
+	mux *http.ServeMux
+}
+
+type APIHandlerFunc func(r *http.Request) ([]byte, int)
+
+func (x APIHandlerFunc) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	defer r.Body.Close()
+	body, code := logRequest(x, r)
+	w.WriteHeader(code)
+	w.Write(body)
+}
+
+func logRequest(handlerFunc APIHandlerFunc, r *http.Request) ([]byte, int) {
+	start := time.Now()
+	body, code := handlerFunc(r)
+
+	clientIP := strings.Split(r.Header.Get("X-Forwarded-For"), ",")[0]
+	if clientIP == "" {
+		clientIP, _, _ = net.SplitHostPort(r.RemoteAddr)
+	}
+	fields := logrus.Fields{
+		"client_ip":       clientIP,
+		"request_path":    r.URL.EscapedPath(),
+		"user_agent":      r.UserAgent(),
+		"request_method":  r.Method,
+		"request_size":    r.ContentLength,
+		"request_seconds": time.Since(start).Seconds(),
+		"status_code":     code,
+	}
+	if code < 500 {
+		logger.WithFields(fields).Info("request completed")
```

Comment:
So as I understand this, a 404 is considered 'request completed', as is bad auth and all other types of 4XX client errors? 

Also, a generic 500 error wouldn't pass this test. To put that another way: `500 < 500 == false`

Reply:
Ha oops true

Reply:
Also, this is to distinguish server ever vs user error.

Reply:
Actually, the logic does what I want. Any status code 500 and up are server errors. Status codes less than 500 typically aren't treated as errors.

Reply:
Still worth logging 400s, because that means we screwed up the client side somewhere. Log a `"request invalid"`, perhaps.

Reply:
This line is still an issue if your want to log 4xx errors as 'request complete'. 500 is not less than 500. 500 is a valid server error code.

Reply:
I don't understand... I'm not trying to log 500 as `request complete`. 500 should be logged as `request failed`.

Reply:
I'm proposing a logging 4xx separately.

Reply:
I think I misread this first time around. I had thought there a boundary condition around 500 errors themselves, but on closer inspection you're handling it correctly.